File: C++/5. LongestPalindromicSubstring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        const int n = s.size();
        int max{};
        string ans;
        if(n == 1) return s;
        for(int i = 0; i < n; i++){
            int x = i, y = i;
            int count = 1;
            while(x > 0 && y < n - 1 && s[x - 1] == s[y + 1]){
                x--;
                y++;
                count += 2;
            }
            if(count > max){
                ans = s.substr(x, count);
                max = count;
            } 
            if(i < n - 1 && s[i] == s[i + 1]){
                int x = i, y = i + 1;
                count = 2;
                while (x > 0 && y < n - 1 && s[x - 1] == s[y + 1]){
                    x--;
                    y++;
                    count += 2;
                }
                if(count > max)
                {
                    ans = s.substr(x, count);
                    max = count;
                }
            }
        }
        return ans;
    }
};

File: Java/5.LongestPalindromicSubstring.java
/*
Given a Given a string s, return the longest palindromic substring in s.

Example TestCase ->
Input: s = "babad"
Output: "bab"
Note: "aba" is also a valid answer.
 */

class Solution{
    public String longestPalindrome(String s) {
        int n = s.length();
        if(n == 0) return "";
        
        boolean[][] dp = new boolean[n][n];
        
        int maxSoFar = 0;
        String ans = "";
        
        for(int gap = 0;gap<n;gap++){
            int sp = 0; int ep = gap;
            while(ep<n){
                char ch1 = s.charAt(sp);
                char ch2 = s.charAt(ep);
                
                if(gap == 0){
                    dp[sp][ep] = true;
                    maxSoFar = 1; ans = ch1+"";
                }else if(gap == 1 && ch1 == ch2){
                    dp[sp][ep] = true;
                    maxSoFar = 1; ans = ""+ch1+ch1;
                }else if(ch1 == ch2 && dp[sp+1][ep-1]){
                    dp[sp][ep] = true;
                    if(ep-sp+1 > maxSoFar){
                        maxSoFar = ep-sp+1;
                        ans = s.substring(sp,ep+1);
                    }
                }
                sp++; ep++;
            }
        }
        return ans;
    }
}
